main.nf

```diff
@@ -0,0 +1,581 @@
+  set file(fasta),file("${fasta}.fai"),file("${fasta}.gz"),file("${fasta}.gz.fai"), file("${fasta}.gz.gzi") into fastaChannel
+  script:
+  """
+  [[ "${params.fai}"=="nofai" ]] &&  samtools faidx $fasta || echo " fai file of user is used, not created"
+  [[ "${params.fastagz}"=="nofastagz" ]]  && bgzip -c ${fasta} > ${fasta}.gz || echo "fasta.gz file of user is used, not created "
+  [[ "${params.gzi}"=="nogzi" ]] && bgzip -c -i ${fasta} > ${fasta}.gz || echo "gzi file of user is used, not created"
+  [[ "${params.gzfai}"=="nogzfai" ]] && samtools faidx "${fasta}.gz" || echo "gz.fai file of user is used, not created"
+  """
+
+}
+
+
+/********************************************************************
+  process preprocessBAM
+  If the user gives the index files for the bam files as an input, they are used
+  If not they are produced in this process given only the fasta file.
+  Moreover this takes care of the read group line too.
+********************************************************************/
+
+
+process preprocessBAM{
+
+
+  tag "${bam[0]}"
+  publishDir "$baseDir/sampleDerivatives"
+
+  input:
+  set val(prefix), file(bam) from bamChannel
+  output:
+  set file("ready/${bam[0]}"), file("ready/${bam[0]}.bai") into completeChannel
+  script:
+  """
+	  mkdir ready
+  [[ `samtools view -H ${bam[0]} | grep '@RG' | wc -l`   > 0 ]] && { mv $bam ready;}|| { java -jar /picard.jar AddOrReplaceReadGroups \
+    I=${bam[0]} \
+    O=ready/${bam[0]} \
+    RGID=${params.rgid} \
+    RGLB=${params.rglb} \
+    RGPL=${params.rgpl} \
+    RGPU=${params.rgpu} \
+    RGSM=${params.rgsm};}
+    cd ready ;samtools index ${bam[0]};
+  """
+}
+
+
+
+fastaChannel.map{file -> tuple (1,file[0],file[1],file[2],file[3],file[4])}
+            .set{all_fa};
+
+completeChannel.map { file -> tuple(1,file[0],file[1]) }
+               .set{all_bam};
+
+all_fa.cross(all_bam)
+      .set{all_fa_bam};
+
+
+
+      /********************************************************************
+        process makeExamples
+        Getting bam files and converting them to images ( named examples )
+
+	Can be parallelized through the params.n_shards
+	( if params.n_shards >= 1 parallelization happens automatically)
+      ********************************************************************/
+
+if(params.bed){
+  process makeExamples_with_bed{
+
+      tag "${bam[1]}"
+    cpus params.j
+
+    input:
+      set file(fasta), file(bam) from all_fa_bam
+      file bedfile from bedfile
+    output:
+      set file("${fasta[1]}"),file("${fasta[1]}.fai"),file("${fasta[1]}.gz"),file("${fasta[1]}.gz.fai"), file("${fasta[1]}.gz.gzi"),val("${bam[1]}"), file("shardedExamples") into examples
+    shell:
+    '''
+      mkdir shardedExamples
+      time seq 0 !{numberShardsMinusOne} | \
+      parallel --eta --halt 2 \
+        python /opt/conda/pkgs/deepvariant-0.7.0-py27h5d9141f_0/share/deepvariant-0.7.0-0/binaries/DeepVariant/0.7.0/DeepVariant-0.7.0+cl-208818123/make_examples.zip \
```

Comment:
Please don't use paths like this. Not entirely sure what this file is that you're calling, but it should be available on the `PATH` if using conda. 

Currently this code will _only_ work in containers and will need to be updated every time there's a version change.

Reply:
Hmmm,

The file is not available on the`PATH` using conda. 

 Here is the `PATH` within the docker container
```bash
echo $PATH
/opt/conda/bin:/opt/conda/envs/nf-core-deepvariant-1.0dev/bin:/opt/conda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin
```

I could use `ENV` to set the location of the file within the docker container. However, this still does not solve the problem of when the container is updated.
<br />
Unless I set the path as something like this:
```bash
/opt/conda/pkgs/deepvariant-*/share/deepvariant-*/binaries/DeepVariant/*/DeepVariant-*/make_examples.zip
```
Though I'm not sure if it will work and I know that nf-core is a fan of the minimal Docker/Singularity container template 😉 
What are your thoughts on this?

Thanks in advance

Reply:
What is the `make_examples.zip` file? I guess I should probably figure out what this process is actually doing first.

It looks like the bioconda distribution provides a script which is running basically the same command though: https://github.com/bioconda/bioconda-recipes/blob/master/recipes/deepvariant/dv_make_examples.py#L50 - maybe we can use that? Then we're still dependent on the conda recipe, but at least it may in theory be possible to run outside of a container.

Reply:
Okay thanks, I've changed the command for `call_variants` and `postprocess_variants` to the shorter executable, so they should be find if deepvariant is updated.

There was a problem with `make_examples` however as I got an error when I specified the `--logdir` and by using the wrapper script `make_examples.zip` you have to specify the `--regions` which is find for exome samples but not WGS.